Training: Add Fine-Tune API Docs

andreyvelich opened the pull request:
Related: https://github.com/kubeflow/training-operator/issues/2013
This is draft PR for our new Fine-Tune API in Kubeflow Training Operator.
We will work on the page structure in this Google doc to finalise it: https://docs.google.com/document/d/18PuuaDRISj5mlrBn1GJrxwuB6Z5zTtXKpVbLUIeLx-8/edit?usp=sharing.

andreyvelich commented:
I added content from the Google doc and one tutorial.
Please let me know what do you think.
/assign @StefanoFioravanzo @kubeflow/wg-training-leads @deepanker13 @kuizhiqing

andreyvelich commented:
/hold for review

hbelmiro reviewed (commented):
/lgtm

StefanoFioravanzo commented on content/en/docs/components/training/user-guides/fine-tuning.md:
```diff
@@ -0,0 +1,172 @@
++++
+title = "How to Fine-Tune LLM with Kubeflow"
```
```suggestion
title = "How to Fine-Tune LLMs with Kubeflow"
```

StefanoFioravanzo commented on content/en/docs/components/training/user-guides/fine-tuning.md:
```diff
@@ -0,0 +1,172 @@
+
+[Training Operator Python SDK](/docs/components/training/installation/#installing-training-python-sdk)
+implements a [`train` Python API](https://github.com/kubeflow/training-operator/blob/6ce4d57d699a76c3d043917bd0902c931f14080f/sdk/python/kubeflow/training/api/training_client.py#L112)
+that simplify ability to fine-tune LLMs with distributed PyTorchJob workers.
```
```suggestion
that simplifies the ability to fine-tune LLMs with distributed PyTorchJob workers.
```

StefanoFioravanzo commented on content/en/docs/components/training/user-guides/fine-tuning.md:
```diff
@@ -0,0 +1,172 @@
+implements a [`train` Python API](https://github.com/kubeflow/training-operator/blob/6ce4d57d699a76c3d043917bd0902c931f14080f/sdk/python/kubeflow/training/api/training_client.py#L112)
+that simplify ability to fine-tune LLMs with distributed PyTorchJob workers.
+
+You need to provide the following parameters to use `train` API:
```
```suggestion
You need to provide the following parameters to use the `train` API:
```

StefanoFioravanzo commented on content/en/docs/components/training/user-guides/fine-tuning.md:
````diff
@@ -0,0 +1,172 @@
+)
+```
+
+After you execute `train` API, Training Operator will orchestrate appropriate PyTorchJob resources
````
```suggestion
After you execute `train`, Training Operator will orchestrate appropriate PyTorchJob resources
```

StefanoFioravanzo commented on content/en/docs/components/training/user-guides/fine-tuning.md:
````diff
@@ -0,0 +1,172 @@
+For example, you can use `train` API as follows to fine-tune BERT model using Yelp Review dataset
+from HuggingFace Hub:
+
+```python
````
If I copy paste this snippet into a notebook, does it run seamlessly? What are the required dependencies? Do we need to provide a `pip install` command to make sure that this snippet runs? Also, what is the expected output?

andreyvelich replied:
Let me add the prerequisites to run this API.

StefanoFioravanzo commented on content/en/docs/components/training/user-guides/fine-tuning.md:
```diff
@@ -0,0 +1,172 @@
+After you execute `train` API, Training Operator will orchestrate appropriate PyTorchJob resources
+to fine-tune LLM.
+
+## Architecture
```
This should go to a "Reference"

StefanoFioravanzo commented on content/en/docs/components/training/user-guides/fine-tuning.md:
```diff
@@ -0,0 +1,172 @@
+You can implement your own trainer for other ML use-cases such as image classification,
+voice recognition, etc.
+
+## User Value for this Feature
```
I think we can just fold this under `Why Training Operator Fine-Tune API Matter ?` by stripping the title `User Value for this Feature`

StefanoFioravanzo commented on content/en/docs/components/training/user-guides/fine-tuning.md:
```diff
@@ -0,0 +1,172 @@
+image classification, or another ML domain, fine-tuning can drastically improve performance and
+applicability of pre-existing models to new datasets and problems.
+
+## Why Training Operator Fine-Tune API Matter ?
```
I feel like this is out of place here. The how-to guide provides a step by step sequenced guide on how to achieve a very specific task. A how-to guide generally does not provide Reference or Explanation. It seems to me we are writing some paragraphs that would be more suited to an "Explanation" section. This is the fourth content types proposed by Diataxis - see here https://diataxis.fr/explanation/

I can very well see a page under "Explanation" titled "LLM Fine-Tune APIs in Kubeflow" where we discuss why we need it and how it fits into the ecosystem. Basically what you wrote already, plus a little bit of refactoring. WDYT?

StefanoFioravanzo replied:
That's a very good question. In the how to guide, we can have something like: "If you want to learn more about how the fine tune API fit in the Kubeflow ecosystem, head to <...>".

And in the exlanation guide, we can say something like: "Head to <link> for a quick start tutorial on using LLM Fine-tune APIs. Head to <link> for a reference architecture on the control plane implementation"

And generally we can have links to how-tos in tutorials and reference guides. So in general, let's try to link related topics together when it makes sense for a user to follow that train of thought

andreyvelich replied:
Sure, what do you think about it @StefanoFioravanzo ?
https://github.com/kubeflow/website/pull/3718/commits/7d30f127d35cc9a306328ccc4d309b7b66c9c54f

StefanoFioravanzo replied:
Looks great!

andreyvelich commented:
I addressed your comments @StefanoFioravanzo.
Regarding this comment:

> Also, what is the expected output?

How we can show the expected output ? Our LLM trainer doesn't support any output yet: https://github.com/kubeflow/training-operator/blob/master/sdk/python/kubeflow/trainer/hf_llm_training.py#L178, so we need to work in the future to understand how user should consume the fine-tuned model.
E.g. exporting to S3 or to other storage.
cc @johnugeorge @deepanker13

StefanoFioravanzo commented:
> so we need to work in the future to understand how user should consume the fine-tuned model.

Issue + KF 1.10 tag? :)

StefanoFioravanzo commented:
@andreyvelich  yes it does thank you!

tenzen-y reviewed (commented):
Awsome documentation! Thank you!
/lgtm

deepanker13 commented:
@andreyvelich the links in fine-tuning.md are giving 404 page not found. Am I missing something?

andreyvelich commented:
> @andreyvelich the links in fine-tuning.md are giving 404 page not found. Am I missing something?

@deepanker13 Did you check these links via Website preview: https://deploy-preview-3718--competent-brattain-de2d6d.netlify.app/ ?

deepanker13 commented:
> > @andreyvelich the links in fine-tuning.md are giving 404 page not found. Am I missing something?
> 
> @deepanker13 Did you check these links via Website preview: https://deploy-preview-3718--competent-brattain-de2d6d.netlify.app/ ?

@andreyvelich it's working with the preview. Thanks for the awesome documentation!
/lgtm

StefanoFioravanzo commented:
@andreyvelich shall we merge this one?
